### PR TITLE
Repo history rewrite

### PR DESCRIPTION
**Pull request by dholms**
Removes history from repos, introduces new sync semantics, and provides an upgrade path to repo v3.

Adds a temporary admin `com.atproto.temp.upgradeRepoVersion` route through which we can upgrade repos at our own pace.

More discussion here: https://github.com/bluesky-social/atproto/discussions/1410
Blog post with discussion of changes here: https://atproto.com/blog/repo-sync-update

---
Note for reviewing: I would review this in in the following order
- repo package
- pds updates
- the migration route + tests

### CODE REVIEW

File: lexicons/com/atproto/sync/getHead.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "defs": {
     "main": {
       "type": "query",
-      "description": "Gets the current HEAD CID of a repo.",
+      "description": "DEPRACATED - please use com.atproto.sync.getRoot instead",
```
Comment (devinivy) [inline lexicons/com/atproto/sync/getHead.json:7]:
```suggestion
      "description": "DEPRECATED - please use com.atproto.sync.getRoot instead",
```

File: lexicons/com/atproto/sync/getRoot.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
+  "lexicon": 1,
+  "id": "com.atproto.sync.getRoot",
```
Comment (bnewbold) [inline lexicons/com/atproto/sync/getRoot.json:3]: are we going to switch this to `com.atproto.sync.getLatestCommit` as discussed? I think `getRoot` is ambiguous; "root" also refers to the top node in the MST (vs. the CID of the commit)

File: lexicons/com/atproto/sync/getRoot.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          }
+        }
+      },
+      "output": {
+        "encoding": "application/json",
+        "schema": {
+          "type": "object",
+          "required": ["root"],
+          "properties": {
+            "root": {"type": "string", "format": "cid"}
```
Comment (bnewbold) [inline lexicons/com/atproto/sync/getRoot.json:25]: would also change this from `root` to `commit`, as this is the CID of the signed commit, not the top node in the MST

File: lexicons/com/atproto/temp/upgradeRepoVersion.json [hunk 4]
```diff
@@ -0,0 +1,20 @@
+{
+  "lexicon": 1,
+  "id": "com.atproto.temp.upgradeRepoVersion",
```
Comment (bnewbold) [inline lexicons/com/atproto/temp/upgradeRepoVersion.json:3]: not `com.atproto.unspecced.*`? not a big deal or worth changing, just surprised.

File: packages/repo/src/types.ts [hunk 5]
```diff
@@ -3,32 +3,48 @@ import { schema as common, def as commonDef } from '@atproto/common'
...
+import CidSet from './cid-set'
 
 // Repo nodes
 // ---------------
 
 const unsignedCommit = z.object({
   did: z.string(),
-  version: z.number(),
-  prev: common.cid.nullable(),
+  version: z.literal(3),
```
Comment (bnewbold) [inline packages/repo/src/types.ts:13]: ah, we are doing the full (minor) breaking repo version change in this PR.

it is kind of complex and unexpected to me that we will have a time period with both v2 and v3 repos floating around. I guess not that big of a deal b/c they are pseudo-compatible.

File: lexicons/com/atproto/temp/upgradeRepoVersion.json (hunk 4 above)
Comment (dholms) [reply lexicons/com/atproto/temp/upgradeRepoVersion.json:3]: The proximate reason is that our lexgen code doesn't had some difficulty with the `unspecced` namespace being defined under both `com.atproto` & `app.bsky`

The post-hoc rationale is that `unspecced` is normally a proof of concept for some feature that we do plan to introduce. This route should only exist for a day or two.

File: packages/repo/src/types.ts (hunk 5 above)
Comment (dholms) [reply packages/repo/src/types.ts:13]: That's basically my thought^^

The alternative is to take on a decent amount of downtime to migrate all of the repos :/

File: lexicons/com/atproto/sync/getRoot.json (hunk 3 above)
Comment (dholms) [reply lexicons/com/atproto/sync/getRoot.json]: I decided to change it to `cid` 👍

File: lexicons/com/atproto/sync/getRoot.json (hunk 2 above)
Comment (dholms) [reply lexicons/com/atproto/sync/getRoot.json]: tried on `getCurrent` for size but then remembered this suggestion. It's a bit wordy for my taste but I think it probably is the most descriptive so I'll probably change it 👍

File: packages/pds/src/api/com/atproto/upgradeRepoVersion.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            .set({ repoRev: rev })
+            .where('creator', '=', did)
+            .where('cid', 'in', cidStrs)
+            .execute()
+        }
+        await dbTxn.db
+          .deleteFrom('ipld_block')
+          .where('creator', '=', did)
+          .where((qb) =>
+            qb.where('repoRev', 'is', null).orWhere('repoRev', '!=', rev),
```
Comment (devinivy) [inline packages/pds/src/api/com/atproto/upgradeRepoVersion.ts:70]: Good call with the `null` check 👀

File: packages/pds/src/sql-repo-storage.ts [hunk 7]
```diff
@@ -143,205 +150,108 @@ export class SqlRepoStorage extends RepoStorage {
...
+    } else {
+      await this.db.db
+        .insertInto('repo_root')
+        .values({
+          did: this.did,
+          root: cid.toString(),
+          indexedAt: this.getTimestamp(),
+        })
+        .onConflict((oc) =>
+          oc.column('did').doUpdateSet({ root: cid.toString() }),
```
Comment (devinivy) [inline packages/pds/src/sql-repo-storage.ts:194]: Small thing, but do we want to update `indexedAt` too?

File: packages/pds/src/sql-repo-storage.ts [hunk 8]
```diff
@@ -143,205 +150,108 @@ export class SqlRepoStorage extends RepoStorage {
...
-        'ipld_block.cid',
-        'ipld_block.content',
-      ])
-      .execute()
+    if (rev) {
+      builder = builder.where('repoRev', '>=', rev)
+    }
+    if (cidCursor) {
+      builder = builder.where('cid', '>', cidCursor.toString())
+    }
```
Comment (devinivy) [inline packages/pds/src/sql-repo-storage.ts:238]: I see the new index on `(creator, repoRev, cid)`.  May not be able to use that if specifying `cidCursor` without `rev`.  Seem alright?

File: packages/pds/src/sql-repo-storage.ts [hunk 9]
```diff
@@ -143,205 +150,108 @@ export class SqlRepoStorage extends RepoStorage {
...
-          .onRef('ipld_block.creator', '=', 'repo_commit_block.creator'),
-      )
-      .select([
-        'repo_commit_block.commit',
-        'ipld_block.cid',
-        'ipld_block.content',
-      ])
-      .execute()
+    if (rev) {
+      builder = builder.where('repoRev', '>=', rev)
```
Comment (devinivy) [inline packages/pds/src/sql-repo-storage.ts:234]: At a glance I assumed `rev` would be supplied when you have that revision on hand, and you need to catch-up w/ blocks that come after `rev`.  But I noticed it's inclusive (`>=`) here.  What's the right way to think about the `rev` param?

File: packages/pds/tests/migrations/repo-version-upgrade.test.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+  const addCruft = async (did: string) => {
+    const cruft: IpldBlock[] = []
+    for (let i = 0; i < 1000; i++) {
+      const bytes = randomBytes(128)
+      const cid = await cidForCbor(bytes)
+      cruft.push({
+        cid: cid.toString(),
+        creator: did,
+        repoRev: null,
```
Comment (devinivy) [inline packages/pds/tests/migrations/repo-version-upgrade.test.ts:80]: Could some `repoRev`s be set here?

File: packages/pds/tests/migrations/repo-version-upgrade.test.ts [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+      .selectAll()
+      .executeTakeFirst()
+    if (!root || !root.rev) {
+      throw new Error('did not set rev')
+    }
+    expect(root.root).toEqual(aliceRepoAfter.commit.cid.toString())
+    const nonUpgradedRecords = await db.db
+      .selectFrom('record')
+      .where('did', '=', alice)
+      .where('repoRev', '!=', root.rev)
```
Comment (devinivy) [inline packages/pds/tests/migrations/repo-version-upgrade.test.ts:148]: Might we want an `.orWhere('repoRev', 'is', null)` on some of these?

File: lexicons/com/atproto/sync/getLatestCommit.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        }
+      },
+      "output": {
+        "encoding": "application/json",
+        "schema": {
+          "type": "object",
+          "required": ["cid", "rev"],
+          "properties": {
+            "cid": {"type": "string", "format": "cid"},
+            "rev": {"type": "string"}
```
[the remark this answers is not in the document]
Comment (dholms) [reply lexicons/com/atproto/sync/getLatestCommit.json:26]: i think so 👍 

especially now that there's a place where a value can _only_ be a TID (where is rkeys can be other strings as well)

File: lexicons/com/atproto/sync/subscribeRepos.json [hunk 13]
```diff
@@ -33,15 +33,23 @@
...
-      "nullable": ["prev"],
+      "required": ["seq", "rebase", "tooBig", "repo", "commit", "rev", "since", "blocks", "ops", "blobs", "time"],
+      "nullable": ["prev", "since"],
       "properties": {
         "seq": {"type": "integer"},
         "rebase": {"type": "boolean"},
         "tooBig": {"type": "boolean"},
         "repo": {"type": "string", "format": "did"},
         "commit": {"type": "cid-link"},
         "prev": {"type": "cid-link"},
```
[the remark this answers is not in the document]
Comment (dholms) [reply lexicons/com/atproto/sync/subscribeRepos.json:44]: yup that's the thought 👍

File: packages/pds/src/api/com/atproto/sync/listBlobs.ts [hunk 14]
```diff
@@ -19,23 +17,9 @@ export default function (server: Server, ctx: AppContext) {
...
-      const commitPath = await storage.getCommitPath(latest, earliest)
-      if (commitPath === null) {
-        throw new InvalidRequestError(
-          `Could not find a valid commit path from ${latest.toString()} to ${earliest?.toString()}`,
-        )
-      }
       const cids = await ctx.services
         .repo(ctx.db)
-        .blobs.listForCommits(did, commitPath)
+        .blobs.listSinceRev(did, params.rev)
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/pds/src/api/com/atproto/sync/listBlobs.ts:22]: hmm ya it probably oughta be paginated 🤔

File: packages/repo/src/data-diff.ts [hunk 15]
```diff
@@ -1,76 +1,111 @@
...
-      if (this.adds[del.key]) {
-        delete this.adds[del.key]
-      } else {
-        delete this.updates[del.key]
-        this.recordDelete(del.key, del.cid)
-      }
-    }
-    this.newCids.addSet(diff.newCids)
-  }
+  // addDiff(diff: DataDiff) {
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/data-diff.ts:81]: ah good catch 👍 

I don't think we need it anymore

File: packages/repo/src/repo.ts [hunk 16]
```diff
@@ -131,47 +134,50 @@ export class Repo extends ReadableRepo {
...
     )
-    const commitCid = await commitBlocks.add(commit)
+    const commitCid = await newBlocks.add(commit)
+
+    // ensure the commit cid actually changed
+    if (commitCid.equals(this.cid)) {
+      newBlocks.delete(commitCid)
+    } else {
+      removedCids.add(this.cid)
+    }
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/repo.ts:166]: yeah the commit CID would necessarily change because it contained the `prev` commit CID in the hash

but also we're just being more exact about block tracking in a way that we weren't previously

File: packages/repo/src/sync/consumer.ts [hunk 17]
```diff
@@ -1,140 +1,194 @@
...
-    persistUpdates(storage, updateStorage, updates),
-    storage.updateHead(root, null),
-  ])
-
-  const repo = await Repo.load(storage, root)
-
+  const diff = await DataDiff.of(updated.data, repo?.data ?? null)
+  const writes = await util.diffToWriteDescripts(diff, updateBlocks)
+  const newBlocks = diff.newMstBlocks
+  const leaves = await updateBlocks.getMany(diff.newLeafCids.toList())
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/sync/consumer.ts:65]: yup nice catch 👍

File: packages/repo/src/types.ts [hunk 18]
```diff
@@ -3,32 +3,48 @@ import { schema as common, def as commonDef } from '@atproto/common'
...
+  did: z.string(),
+  version: z.literal(2),
+  data: common.cid,
+  rev: z.string().optional(),
+  prev: common.cid.nullable(),
+  sig: common.bytes,
+})
+export type LegacyV2Commit = z.infer<typeof legacyV2Commit>
+
+const versionedCommit = z.union([commit, legacyV2Commit])
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/types.ts:40]: nice 👍

File: packages/pds/src/sql-repo-storage.ts (hunk 7 above)
Comment (dholms) [reply packages/pds/src/sql-repo-storage.ts:194]: ya good call 👍

File: packages/pds/src/sql-repo-storage.ts (hunk 9 above)
Comment (dholms) [reply packages/pds/src/sql-repo-storage.ts:234]: yup you're "at a glance" assumption was correct. this should be an exclusive `>` 👍

File: packages/pds/tests/migrations/repo-version-upgrade.test.ts (hunk 10 above)
Comment (dholms) [reply packages/pds/tests/migrations/repo-version-upgrade.test.ts:80]: yup

hmmm yup - i can add some in 👍

File: packages/pds/tests/migrations/repo-version-upgrade.test.ts (hunk 11 above)
Comment (dholms) [reply packages/pds/tests/migrations/repo-version-upgrade.test.ts:148]: yup yup 👍

File: packages/pds/src/sql-repo-storage.ts (hunk 9 above)
Comment (dholms) [reply packages/pds/src/sql-repo-storage.ts]: ah wait i remember why this was the case...

for the initial `since` it's exclusive, but when paginating, it's inclusive

i'll refactor to use a real cursor

File: packages/pds/src/api/com/atproto/sync/listBlobs.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (dholms) [reply packages/pds/src/api/com/atproto/sync/listBlobs.ts]: paginated 👍

File: packages/pds/src/sql-repo-storage.ts [hunk 19]
```diff
@@ -191,56 +191,65 @@ export class SqlRepoStorage extends ReadableBlockstore implements RepoStorage {
...
+      .limit(500)
+    if (cursor) {
+      builder = builder.where((qb) =>
+        qb
+          .where('repoRev', '>', cursor.rev)
+          .orWhere((inner) =>
+            inner
+              .where('repoRev', '=', cursor.rev)
+              .where('cid', '>', cursor.cid.toString()),
+          ),
```
Comment (devinivy) [inline packages/pds/src/sql-repo-storage.ts:247]: Would recommend using this pattern to ensure you hit the index: `("repoRev", "cid") > (${cursor.rev}, ${cursor.cid})`.  Shouldn't really be necessary, but we've seen it miss the index the other way for sure 🙃

**Review by devinivy — APPROVED** [review]

File: packages/pds/src/sql-repo-storage.ts (hunk 19 above)
Comment (dholms) [reply packages/pds/src/sql-repo-storage.ts:247]: ah good looks 👍  simpler too

**Review by devinivy — APPROVED** [review]
